pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+                                right array
+    %(klass)s.from_breaks : Construct an %(klass)s from an array of
+                                splits
+    """
+
+    @classmethod
+    @Appender(_interval_shared_docs['from_tuples'] % _shared_docs_kwargs)
+    def from_tuples(cls, data, closed='right', copy=False, dtype=None):
+        if len(data):
+            left, right = [], []
+        else:
+            left = right = data
+
+        for d in data:
+            if isna(d):
+                lhs = rhs = np.nan
+            else:
+                name = cls.__name__
+                try:
+                    # need list of length 2 tuples, e.g. [(0, 1), (1, 2), ...]
+                    lhs, rhs = d
+                except ValueError:
+                    msg = ('{name}.from_tuples requires tuples of '
+                           'length 2, got {tpl}').format(name=name, tpl=d)
+                    raise ValueError(msg)
+                except TypeError:
+                    msg = ('{name}.from_tuples received an invalid '
+                           'item, {tpl}').format(name=name, tpl=d)
+                    raise TypeError(msg)
+                lhs, rhs = d
+            left.append(lhs)
+            right.append(rhs)
+
+        return cls.from_arrays(left, right, closed, copy=False,
+                               dtype=dtype)
+
+    def _validate(self):
+        """Verify that the IntervalIndex is valid.
+
+        Checks that
+
+        * closed is valid
+        * left and right match lengths
+        * left and right have the same missing values
+        * left is always below right
+        """
+        if self.closed not in _VALID_CLOSED:
+            raise ValueError("invalid option for 'closed': {closed}"
+                             .format(closed=self.closed))
+        if len(self.left) != len(self.right):
+            raise ValueError('left and right must have the same length')
+        left_mask = notna(self.left)
+        right_mask = notna(self.right)
+        if not (left_mask == right_mask).all():
+            raise ValueError('missing values must be missing in the same '
+                             'location both left and right sides')
+        if not (self.left[left_mask] <= self.right[left_mask]).all():
+            raise ValueError('left side of interval must be <= right side')
+
+    # ---------
+    # Interface
+    # ---------
+    def __iter__(self):
+        return iter(self.values)
+
+    def __len__(self):
+        return len(self.left)
+
+    def __getitem__(self, value):
+        mask = self.isna()[value]
+        if is_scalar(mask) and mask:
+            return self._fill_value
+
+        left = self.left[value]
+        right = self.right[value]
+
+        # scalar
+        if not isinstance(left, Index):
+            return Interval(left, right, self.closed)
+
+        return self._shallow_copy(left, right)
+
+    def __setitem__(self, key, value):
```

Comment:
Should this support the case where `value` is `np.nan`?  Doesn't look like that currently works.

Reply:
Yes, it currently fails. I'll see what can be done.

Reply:
Looks like there's also something strange going on with datetime data:
```python-traceback
In [2]: ia = pd.interval_range(pd.Timestamp('20180101'), periods=3).values

In [3]: ia
Out[3]:
IntervalArray([(2018-01-01, 2018-01-02], (2018-01-02, 2018-01-03], (2018-01-03, 2018-01-04]],
              closed='right',
              dtype='interval[datetime64[ns]]')

In [4]: new_iv = pd.Interval(pd.Timestamp('20180101'), pd.Timestamp('20180105'))

In [5]: new_iv
Out[5]: Interval('2018-01-01', '2018-01-05', closed='right')

In [6]: ia[0] = new_iv

In [7]: ia
Out[7]: ---------------------------------------------------------------------------
ValueError: left side of interval must be <= right side

In [8]: ia.left
Out[8]: DatetimeIndex(['2018-01-01', '2018-01-05', '2018-01-03'], dtype='datetime64[ns]', freq='D')

In [9]: ia.right
Out[9]: DatetimeIndex(['2018-01-05', '2018-01-03', '2018-01-04'], dtype='datetime64[ns]', freq='D')
```